frontend/src/stimulus/controllers/beforeunload.controller.ts

```diff
@@ -0,0 +1,75 @@
+import { ApplicationController } from 'stimulus-use';
+import { TurboBeforeVisitEvent } from '@hotwired/turbo';
+
+export function setPageWasEdited() {
+  const application = window.Stimulus;
+  const controller = application.getControllerForElementAndIdentifier(document.body, 'beforeunload') as BeforeunloadController;
+  if (controller) {
+    controller.pageWasEdited = true;
+  }
+}
+
+export function setPageWasSubmitted() {
+  const application = window.Stimulus;
+  const controller = application.getControllerForElementAndIdentifier(document.body, 'beforeunload') as BeforeunloadController;
+  if (controller) {
+    controller.pageWasEdited = true;
+  }
+}
+
+export class BeforeunloadController extends ApplicationController {
+  /** Globally setable variable whether the page was edited */
+  public pageWasEdited = false;
+
+  /** Globally setable variable whether the page form is submitted.
+   * Necessary to avoid a data loss warning on beforeunload */
+  public pageIsSubmitted = false;
+
+  private boundBeforeUnloadHandler = this.beforeunloadHandler.bind(this);
+
+  private submitHandler = () => {
+    this.pageIsSubmitted = true;
+  };
+
+  private turboHandler = () => {
+    this.pageWasEdited = false;
+    this.pageIsSubmitted = false;
+  };
+
+  connect() {
+    super.connect();
+
+    window.addEventListener('beforeunload', this.boundBeforeUnloadHandler);
+    document.addEventListener('turbo:before-visit', this.boundBeforeUnloadHandler);
+    document.addEventListener('turbo:submit-end', this.turboHandler);
+    document.addEventListener('submit', this.submitHandler);
+  }
+
+  disconnect() {
+    window.removeEventListener('beforeunload', this.boundBeforeUnloadHandler);
+    document.removeEventListener('turbo:before-visit', this.boundBeforeUnloadHandler);
```

Comment:
Per the above comment, we can `AbortController` to remove the listeners.

```suggestion
    this.abortController?.abort();
```

I really like this pattern - it's used extensively for Primer components: https://github.com/primer/view_components/blob/main/app/components/primer/focus_group.ts#L42 (there's also [this article](https://gomakethings.com/the-handleevent-method-is-the-absolute-best-way-to-handle-events-in-web-components/), which although written with Web Components in mind, the same could apply to Stimulus controllers that bind their own event listeners).